File: Viralt.Domain/Services/PrizeService.cs
using Viralt.Domain.Interfaces.Services;
using Viralt.Domain.Models;
using Viralt.DTO.Campaign;
using Viralt.Infra.Interfaces.Repository;

namespace Viralt.Domain.Services;

public class PrizeService : IPrizeService
{
    private readonly IPrizeRepository<Prize> _repository;

    public PrizeService(IPrizeRepository<Prize> repository)
    {
        _repository = repository;
    }

    public PrizeInfo GetById(long prizeId)
    {
        var model = _repository.GetById(prizeId);
        return MapToDto(model);
    }

    public IEnumerable<PrizeInfo> ListByCampaign(long campaignId)
    {
        return _repository.ListByCampaign(campaignId).Select(MapToDto);
    }

    public PrizeInfo Insert(PrizeInsertInfo dto)
    {
        var model = MapToModel(dto);
        var saved = _repository.Insert(model);
        return MapToDto(saved);
    }

    public PrizeInfo Update(PrizeUpdateInfo dto)
    {
        var model = MapToModel(dto);
        var updated = _repository.Update(model);
        return MapToDto(updated);
    }

    public void Delete(long prizeId)
    {
        _repository.Delete(prizeId);
    }

    private static PrizeInfo MapToDto(Prize model)
    {
        if (model == null) return null;
        return new PrizeInfo
        {
            PrizeId = model.PrizeId,
            CampaignId = model.CampaignId,
            Title = model.Title,
            Description = model.Description,
            Image = model.Image,
            Quantity = model.Quantity,
            PrizeType = model.PrizeType.ToString(),
            CouponCode = model.CouponCode,
            SortOrder = model.SortOrder,
            MinEntriesRequired = model.MinEntriesRequired ?? 0
        };
    }

    private static Prize MapToModel(PrizeInsertInfo dto) => new()
    {
        CampaignId = dto.CampaignId,
        Title = dto.Title,
        Description = dto.Description,
        Image = dto.Image,
        Quantity = dto.Quantity,
        PrizeType = int.TryParse(dto.PrizeType, out var pt) ? pt : 0,
        CouponCode = dto.CouponCode,
        SortOrder = dto.SortOrder,
        MinEntriesRequired = dto.MinEntriesRequired
    };

    private static Prize MapToModel(PrizeUpdateInfo dto)
    {
        var model = MapToModel((PrizeInsertInfo)dto);
        model.PrizeId = dto.PrizeId;
        return model;
    }
}

File: viralt-app/src/Services/prizeService.ts
import type {
  PrizeInfo, PrizeInsertInfo, PrizeUpdateInfo,
  PrizeListResult, PrizeGetResult, PrizeStatusResult,
} from '../types/prize';
import { getApiUrl, getHeaders } from './apiHelpers';
import { graphqlQuery, GRAPHQL_ADMIN } from './graphqlClient';

const API_BASE = `${getApiUrl()}/api/Prize`;

interface PrizeServiceConfig {
  onUnauthorized?: () => void;
}

/** Prize fields fragment for GraphQL queries */
const PRIZE_FIELDS = `
  prizeId
  campaignId
  title
  description
  image
  quantity
  prizeType
  couponCode
  sortOrder
  minEntriesRequired
`;

/** Prize Service — Manages all API operations related to prizes */
class PrizeService {
  private config: PrizeServiceConfig;

  constructor(config: PrizeServiceConfig = {}) {
    this.config = config;
  }

  private async handleResponse<T>(response: Response): Promise<T> {
    if (response.status === 401) {
      this.config.onUnauthorized?.();
      throw new Error('Unauthorized');
    }
    if (!response.ok) {
      const error = await response.text();
      throw new Error(error || 'Request failed');
    }
    return response.json();
  }

  /** List prizes by campaign (GraphQL) */
  async listByCampaign(campaignId: number): Promise<PrizeListResult> {
    const query = `
      query ListPrizesByCampaign($campaignId: Long!) {
        prizesByCampaign(campaignId: $campaignId) {
          ${PRIZE_FIELDS}
        }
      }
    `;
    const data = await graphqlQuery<{ prizesByCampaign: PrizeInfo[] }>(
      GRAPHQL_ADMIN, query, { campaignId }, true
    );
    return { sucesso: true, mensagem: '', erros: null, prizes: data.prizesByCampaign };
  }

  /** Create a new prize (REST) */
  async insert(data: PrizeInsertInfo): Promise<PrizeGetResult> {
    const response = await fetch(`${API_BASE}/insert`, {
      method: 'POST', headers: getHeaders(true), body: JSON.stringify(data),
    });
    return this.handleResponse<PrizeGetResult>(response);
  }

  /** Update an existing prize (REST) */
  async update(data: PrizeUpdateInfo): Promise<PrizeGetResult> {
    const response = await fetch(`${API_BASE}/update`, {
      method: 'POST', headers: getHeaders(true), body: JSON.stringify(data),
    });
    return this.handleResponse<PrizeGetResult>(response);
  }

  /** Delete a prize by ID (REST) */
  async delete(id: number): Promise<PrizeStatusResult> {
    const response = await fetch(`${API_BASE}/delete/${id}`, {
      method: 'DELETE', headers: getHeaders(true),
    });
    return this.handleResponse<PrizeStatusResult>(response);
  }
}

export const prizeService = new PrizeService();
export default PrizeService;
